classes/form/username_form.php

```diff
@@ -0,0 +1,111 @@
+            $username = $frm->username;
+        }
+        $user_attributes = array('placeholder'   => $this->lang_string('form_username_placeholder'),
+                                 'additionalcss' => $this->lang_string('form_username_button_class'),
+                                 'autofocus'     => '',
+                                 'value'         => $username);
+        $mform->addElement('text', 'username', $this->lang_string('form_username_label'), $user_attributes);
+        $mform->setType('username', PARAM_USERNAME);
+
+        $mform->addElement('html', sprintf('<div class="%s">', $this->lang_string('form_username_remusername_class')));
+        $mform->addElement('advcheckbox', 'rememberme', '', $this->lang_string('form_username_remusername_label'), array ('id' => 'check_rememberme'));
+        $mform->addElement('html', '</div>');
+
+        $submit_attributes = array('additionalcss' => $this->lang_string('form_username_button_class'));
+        $mform->addElement('submit', 'submitusername', $this->lang_string('form_username_button_label'), $submit_attributes);
+
+        $mform->addElement('html', sprintf('<div class="%s"><a href="%s">%s</a></div>',
+            $this->lang_string('form_userpass_forgot_class'),
+            new \moodle_url('/local/signin/forgot.php'),
+            $this->lang_string('form_username_forgot_label')));
+    }
+
+    function set_data($default_values)
+    {
+        parent::set_data($default_values); // TODO: Change the autogenerated stub
+    }
+
+    public function lang_string($id) {
+        return get_string($id, util::MOODLE_COMPONENT);
+    }
+
+    /**
+     * Validate the username field.
+     *
+     * @param array $data
+     * @param array $files
+     * @return array|null
+     */
+    public function validation($data, $files) {
+        global $CFG, $PAGE;
+
+        if (!array_key_exists('username', $data)) {
+            return array('username' => $this->lang_string('form_username_not_provided'));
+        }
+
+        $username = $data['username'];
+
+        if (strlen($username) == 0 || !static::active_user_exists($username)) {
+            return array('username' => $this->lang_string('form_username_not_found_valid'));
+        }
+
+        if ($username == 'guest') {
+            return;
+        }
+
+        if (class_exists('bmdisco_domain\brand_domain')) {
+            $correct_domain = brand_domain::get_user_default_domain_obj($username, true) ?
+                brand_domain::get_user_default_domain_obj($username, true)->domain : '';
+            if ($correct_domain) {
+                $current_url = parse_url($CFG->wwwroot, PHP_URL_HOST);
+                if ($current_url !== $correct_domain) {
+                    $url = new \moodle_url(str_replace($current_url, $correct_domain, $PAGE->url), array('username' => $username));
+                    redirect($url);
+                }
+            }
+        }
```

Comment:
Problem: what if `bmdisco_domain` isn't installed on a platform?

You'll report that the user doesn't exist.